File: HW/HW7/HW7.py
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        if not intervals:
            return 0

        intervals.sort(key=lambda x: x[0])  # Sort intervals based on start time
        rooms = []  # Heap to store end times of ongoing meetings
      
      for start, end in intervals:
            if rooms and rooms[0] <= start:
                heappop(rooms)  # Remove the meeting that has ended
            heappush(rooms, end)  # Add the end time of the current meeting

        return len(rooms)

File: HW/HW7/HW7.java
import java.util.Arrays;
import java.util.PriorityQueue;

class Solution {
    public int minMeetingRooms(int[][] intervals) {
        if (intervals == null || intervals.length == 0) {
            return 0;
        }
        
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]); // Sort intervals based on start time
        
        PriorityQueue<Integer> rooms = new PriorityQueue<>();
        rooms.add(intervals[0][1]); // Add the end time of the first meeting
        
        for (int i = 1; i < intervals.length; i++) {
            if (rooms.peek() <= intervals[i][0]) {
                rooms.poll(); // Remove the meeting that has ended
            }
            rooms.add(intervals[i][1]); // Add the end time of the current meeting
        }
        
        return rooms.size();
    }
}
